sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DistinctHash.java

```diff
@@ -3,60 +3,129 @@
     }
 
-    private static UInt128 getHashFromJsonSerializable(JsonSerializable resource) {
-        final ByteBuffer byteBuffer = ModelBridgeInternal.serializeJsonToByteBuffer(resource, OBJECT_MAPPER);
-        final byte[] bytes = byteBuffer.array();
-        return MurmurHash3_128.hash128(bytes, bytes.length);
+    private static UInt128 getHashFromJsonSerializable(JsonSerializable resource, UInt128 seed) throws IOException {
+        resource.populatePropertyBag();
+        return getHash(resource.getPropertyBag(), seed);
     }
 
-    private static UInt128 getHashFromList(List<Object> resource) {
-        UInt128 hash = ARRAY_HASH_SEED;
-        for (Object obj : resource) {
-            if (obj instanceof JsonSerializable) {
-                byte[] bytes = hash.toByteBuffer().array();
-                if (bytes.length == 0) {
-                    throw new IllegalStateException("Failed to hash!");
-                }
-                hash = MurmurHash3_128.hash128(bytes, bytes.length,
-                                               getHashFromJsonSerializable((JsonSerializable) obj));
-            }
+    private static UInt128 getHashFromArrayNode(ArrayNode arrayNode, UInt128 seed) throws IOException {
+        // Start the array with a distinct hash, so that empty array doesn't hash to another value.
+        UInt128 hash = MurmurHash3_128.hash128(HashSeeds.ARRAY, seed);
+
+        for (int i = 0; i < arrayNode.size(); i++) {
+            // Order of array items matter in equality check, so add the index just in case that property does not hold in the future.
+            UInt128 arrayItemSeed = HashSeeds.ARRAY_INDEX.add(i);
+            hash = MurmurHash3_128.hash128(hash, getHash(arrayNode.get(i), arrayItemSeed));
+        }
+        return hash;
+    }
+
+    private static UInt128 getHashFromObjectNode(ObjectNode objectNode, UInt128 seed) throws IOException {
+        UInt128 hash = MurmurHash3_128.hash128(HashSeeds.OBJECT, seed);
+        UInt128 intermediateHash = UInt128.ZERO;
+
+        // Property order should not result in a different hash.
+        // This is consistent with equality comparison.
+        Iterator<Map.Entry<String, JsonNode>> children = objectNode.fields();
+        while ((children.hasNext())) {
+            Map.Entry<String, JsonNode> child = children.next();
+            UInt128 nameHash = MurmurHash3_128.hash128(child.getKey(), HashSeeds.PROPERTY_NAME);
+            UInt128 propertyHash = getHash(child.getValue(), nameHash);
+            intermediateHash = intermediateHash.xor(propertyHash);
+        }
+
+        if (intermediateHash.compareTo(UInt128.ZERO) == 1) {
+            hash = MurmurHash3_128.hash128(intermediateHash, hash);
+        }
+
+        return hash;
+    }
+
+    private static UInt128 getHashFromList(List<Object> resource, UInt128 seed) throws IOException {
+        UInt128 hash = MurmurHash3_128.hash128(HashSeeds.ARRAY, seed);
```

Comment:
I can see a possibility to merge the logic of this function - `getHashFromList` and `getHashFromArrayNode` since the logic is exactly same. 
Or @xinlian12  do you want to keep it separate for more readability ? The only downside of keeping them separate is that if we have a bug in one logic, we could have it in other logic as well and we will have to fix them separately.

Reply:
I combined into one: getHashFromIterator